Model/Product/Index/IndexRepository.php

```diff
@@ -0,0 +1,199 @@
+
+    /**
+     * IndexRepository constructor.
+     *
+     * @param IndexResource $indexResource
+     * @param IndexCollectionFactory $indexCollectionFactory
+     */
+    public function __construct(
+        IndexResource $indexResource,
+        IndexCollectionFactory $indexCollectionFactory
+    ) {
+        $this->indexResource = $indexResource;
+        $this->indexCollectionFactory = $indexCollectionFactory;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getOneByProductAndStore(ProductInterface $product, StoreInterface $store)
+    {
+        /* @var IndexCollection $collection */
+        $collection = $this->indexCollectionFactory->create()
+            ->addFieldToSelect('*')
+            ->addFieldToFilter(
+                NostoIndex::PRODUCT_ID,
+                ['eq' => $product->getId()]
+            )
+            ->addStoreFilter($store)
+            ->setPageSize(1)
+            ->setCurPage(1);
+        return $collection->getFirstItem(); // @codingStandardsIgnoreLine
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getById($id)
+    {
+        $collection = $this->indexCollectionFactory->create()
+            ->addFieldToSelect('*')
+            ->addFieldToFilter(
+                NostoIndex::ID,
+                ['eq' => $id]
+            )
+            ->setPageSize(1)
+            ->setCurPage(1);
+        return $collection->getFirstItem(); // @codingStandardsIgnoreLine
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getTotalOutOfSync(Store $store)
+    {
+        /* @var IndexCollection $collection */
+        $collection = $this->indexCollectionFactory->create();
+        $collection->addFilter(
+            ProductIndexInterface::IN_SYNC,
+            NostoIndex::DB_VALUE_BOOLEAN_FALSE,
+            'eq'
+        );
+        if ((int)$store->getId() !== 0) {
+            $collection->addStoreFilter($store);
+        }
+        return $collection->getSize();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getTotalDirty(Store $store)
+    {
+        /* @var IndexCollection $collection */
+        $collection = $this->indexCollectionFactory->create();
+        $collection->addFilter(
+            ProductIndexInterface::IS_DIRTY,
+            NostoIndex::DB_VALUE_BOOLEAN_TRUE,
+            'eq'
+        );
+        if ((int)$store->getId() !== 0) {
+            $collection->addStoreFilter($store);
+        }
+        return $collection->getSize();
+    }
+    /**
+     * @inheritdoc
+     */
+    public function getByIds(array $ids)
+    {
+        $collection = $this->indexCollectionFactory->create()
+            ->addFieldToSelect('*')
+            ->addIdsFilter($ids)
+            ->setPageSize(1)
+            ->setCurPage(1);
+        return $collection->getItems(); // @codingStandardsIgnoreLine
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getByProductIdAndStoreId(int $productId, int $storeId)
```

Comment:
This method should accept Store and Product

Reply:
We don't want to create always instances for `Product` and `Store`. For example when we get the product id and store id from our own index table.